### PR TITLE
Rholang: Normalizer: Add support for collections.

### PR DESCRIPTION
**Pull request by guardbotmk3**
Adds support for collections to the term normalizer.
Relative to #304

### CODE REVIEW

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/NormalizeTest.scala [hunk 1]
```diff
@@ -4,35 +4,124 @@ import coop.rchain.rholang.syntax.rholang_mercury.Absyn.{Ground => AbsynGround,
...
+    result.par should be (inputs.par.copy(
+      exprs = ETuple(List(
+          Par().copy(exprs = List(EVar(FreeVar(0)))),
+          Par().copy(evals = List(Eval(ChanVar(FreeVar(1))))))
+      ) :: inputs.par.exprs))
+    result.knownFree should be (inputs.knownFree.newBindings(
+        List((Some("Q"), ProcSort), (Some("y"), NameSort)))._1,
+    )
+  }
+  "Tuple" should "propgate free variables" in {
```
Comment (KentShikama) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/NormalizeTest.scala]: propagate

**Comment by guardbotmk3** [comment]
OK, fixed. Re-approve?

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/NormalizeTest.scala [hunk 2]
```diff
@@ -4,35 +4,124 @@ import coop.rchain.rholang.syntax.rholang_mercury.Absyn.{Ground => AbsynGround,
...
+    val tuple = new PCollect(new CollectTuple(tupleData))
+
+    val result = ProcNormalizeMatcher.normalizeMatch(tuple, inputs)
+    result.par should be (inputs.par.copy(
+      exprs = ETuple(List(
+          Par().copy(exprs = List(EVar(FreeVar(0)))),
+          Par().copy(evals = List(Eval(ChanVar(FreeVar(1))))))
+      ) :: inputs.par.exprs))
+    result.knownFree should be (inputs.knownFree.newBindings(
+        List((Some("Q"), ProcSort), (Some("y"), NameSort)))._1,
```
Comment (KentShikama) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/NormalizeTest.scala]: The comma at the end is unnecessary. Travis complains about this in a later PR

https://travis-ci.org/rchain/rchain/jobs/343987729#L1620

Interestingly enough it works for me locally though with the comma - maybe related to how Scala implicitly tuples/untuples?

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/NormalizeTest.scala [hunk 3]
```diff
@@ -4,35 +4,124 @@ import coop.rchain.rholang.syntax.rholang_mercury.Absyn.{Ground => AbsynGround,
...
+    result.par should be (inputs.par.copy(
+      exprs = ESet(List(
+          Par().copy(exprs = List(EPlus(
+              Par().copy(exprs = List(EVar(BoundVar(0)))),
+              Par().copy(exprs = List(EVar(FreeVar(0))))))),
+          Par().copy(exprs = List(GInt(7))),
+          Par().copy(exprs = List(EVar(FreeVar(1)), GInt(8))))
+      ) :: inputs.par.exprs))
+    result.knownFree should be (inputs.knownFree.newBindings(
+        List((Some("R"), ProcSort), (Some("Q"), ProcSort)))._1,
```
Comment (KentShikama) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/NormalizeTest.scala]: Similar to the above.

**Comment by guardbotmk3** [comment]
OK, that issue is fixed. I'm surprised it worked.

**Pull request merged**